source/iterator/snapshot_iterator.go

```diff
@@ -0,0 +1,401 @@
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jpillora/backoff"
+	marketoclient "github.com/rustiever/conduit-connector-marketo/marketo-client"
+	"github.com/rustiever/conduit-connector-marketo/source/position"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// is the maximum number of days between two snapshots. If the gap between two snapshots is greater than this number,
+	// API will return an error. This is limitation of the API.
+	MaximumDaysGap = 744 // 31 days in Hours
+)
+
+// to iterate through the snapshots for specified configs
+type SnapshotIterator struct {
+	endpoint      string
+	fields        []string
+	client        *marketoclient.Client
+	exportID      string
+	initialDate   time.Time
+	iteratorCount int
+	errChan       chan error
+	csvReader     chan *csv.Reader
+	buffer        chan []string
+	lastMaxModied time.Time
+}
+
+// returns NewSnapshotIterator with supplied parameters, also initiates the pull and flush goroutines.
+func NewSnapshotIterator(ctx context.Context, endpoint string, fields []string, client marketoclient.Client, p position.Position) (*SnapshotIterator, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "NewSnapshotIterator").Logger()
+	logger.Trace().Msg("Starting the NewSnapshotIterator")
+
+	var err error
+	s := &SnapshotIterator{
+		endpoint:      endpoint,
+		client:        &client,
+		fields:        fields,
+		errChan:       make(chan error),
+		lastMaxModied: time.Time{},
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	s.initialDate, err = s.getLastProcessedDate(ctx, p)
+	if err != nil {
+		logger.Error().Err(err).Msg("Error getting initial date")
+		return nil, err
+	}
+
+	startDateDuration := time.Since(s.initialDate)
+	s.iteratorCount = int(startDateDuration.Hours()/MaximumDaysGap) + 1
+	logger.Info().Msgf("Preparing %d number of snapshots", s.iteratorCount)
+	s.csvReader = make(chan *csv.Reader, s.iteratorCount)
+	s.buffer = make(chan []string, 10)
+
+	eg.Go(func() error {
+		return s.pull(ctx)
+	})
+
+	eg.Go(func() error {
+		return s.flush(ctx)
+	})
+
+	go func() {
+		err := eg.Wait()
+		logger.Trace().Msg("Errgroup wait finished")
+		if err != nil {
+			logger.Error().Err(err).Msg("Error waiting for errGroup")
+			s.errChan <- err
+		}
+	}()
+
+	return s, nil
+}
+
+func (s *SnapshotIterator) HasNext(ctx context.Context) bool {
+	if s.csvReader == nil && len(s.buffer) == 0 {
+		return false
+	}
+	return true
+}
+
+// retunrs Next record from the iterator's buffer, otherwise returns error.
+// Also returns ErrDone if the iterator is done.
+func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "Next").Logger()
+	logger.Trace().Msg("Starting the Next method")
+
+	select {
+	// case <-ctx.Done():
+	// 	s.Stop(ctx)
+	// 	return sdk.Record{}, ctx.Err()
```

Comment:
Are we ignoring this case intentionally?

Reply:
No, for testing i commented, now i uncommented